File: apps/viewer/src/physics/rod.ts
/**
 * Discrete elastic rod - real-time TypeScript port of the Python reference
 * implementation in `packages/physics-core/src/cathsim_physics/rod.py`.
 *
 * The two implementations share the same discrete energy, the same analytic
 * gradient and the same over-damped implicit time integration, so their
 * results can be compared node-by-node.  `rod.parity.test.ts` replays the
 * golden cases in `tests/reference-data/` against this code; that check is the
 * Python <-> TypeScript counterpart of the Python <-> C++ parity test planned
 * for Milestone 3.
 *
 * Units: mm, s, N, rad. Energy is in N mm.
 *
 * Research prototype - Not for clinical use.
 */

/** Constitutive parameters of the rod. See `RodMaterial` in the Python core. */
export interface RodMaterial {
  /** EI [N mm^2] */
  bendingStiffnessNmm2: number;
  /** GJ [N mm^2] */
  torsionalStiffnessNmm2: number;
  /** EA [N] */
  axialStiffnessN: number;
  /** Translational viscous damping per unit length [N s / mm^2]. */
  dampingNsPerMm2: number;
  /** Torsional viscous damping per unit length [N mm s / rad / mm]. */
  torsionalDampingNmmSPerRad: number;
}

/** Kinematic state: node positions [mm] and one twist angle per edge [rad]. */
export interface RodState {
  /** Flat array of length 3N. */
  nodesMm: Float64Array;
  /** Length N-1. */
  phiRad: Float64Array;
  /** Bishop anchor director of edge 0, length 3. */
  d1First: Float64Array;
}

/** Undeformed configuration. */
export interface RodRest {
  /** Length N-1 [mm]. */
  edgeLengthsMm: Float64Array;
  /** Rest curvature binormal at interior nodes, flat array of length 3(N-2). */
  kb0: Float64Array;
  /** Rest twist increment per interior node, length N-2 [rad]. */
  dphi0Rad: Float64Array;
}

/** Optional external loads. */
export interface ExternalLoads {
  /** Flat 3N array [N]. */
  nodeForcesN?: Float64Array;
  /** Length N-1 [N mm]. */
  edgeTorquesNmm?: Float64Array;
}

export interface EnergyBreakdown {
  stretch: number;
  bend: number;
  twist: number;
  external: number;
  total: number;
}

const EPSILON = 1e-12;

export class RodGeometryError extends Error {}

export function nodeCount(state: RodState): number {
  return state.nodesMm.length / 3;
}

/** Voronoi length of every interior node [mm], length N-2. */
export function voronoiLengths(rest: RodRest): Float64Array {
  const count = rest.edgeLengthsMm.length - 1;
  const out = new Float64Array(count);
  for (let i = 0; i < count; i += 1) {
    out[i] = 0.5 * (rest.edgeLengthsMm[i] + rest.edgeLengthsMm[i + 1]);
  }
  return out;
}

/**
 * Total energy [N mm] and its analytic gradient.
 *
 * `gradNodes` is filled with dE/dx [N] (flat 3N) and `gradPhi` with
 * dE/dphi [N mm] (length N-1). Both are overwritten, not accumulated.
 */
export function energyAndGradient(
  state: RodState,
  rest: RodRest,
  material: RodMaterial,
  loads: ExternalLoads | undefined,
  gradNodes: Float64Array,
  gradPhi: Float64Array,
): EnergyBreakdown {
  const x = state.nodesMm;
  const phi = state.phiRad;
  const n = nodeCount(state);
  const edges = n - 1;

  gradNodes.fill(0);
  gradPhi.fill(0);

  let stretchEnergy = 0;
  let bendEnergy = 0;
  let twistEnergy = 0;
  let externalEnergy = 0;

  // Edge vectors, lengths and unit tangents.
  const edgeVec = new Float64Array(3 * edges);
  const edgeLen = new Float64Array(edges);
  const tangent = new Float64Array(3 * edges);
  for (let j = 0; j < edges; j += 1) {
    const a = 3 * j;
    const b = 3 * (j + 1);
    const ex = x[b] - x[a];
    const ey = x[b + 1] - x[a + 1];
    const ez = x[b + 2] - x[a + 2];
    const len = Math.hypot(ex, ey, ez);
    if (!(len > EPSILON)) {
      throw new RodGeometryError(
        `Two neighbouring rod nodes collapsed at edge ${j} (zero length). ` +
          'Increase the axial stiffness EA or reduce the time step.',
      );
    }
    edgeVec[a] = ex;
    edgeVec[a + 1] = ey;
    edgeVec[a + 2] = ez;
    edgeLen[j] = len;
    tangent[a] = ex / len;
    tangent[a + 1] = ey / len;
    tangent[a + 2] = ez / len;

    // --- axial stretch: E = 1/2 EA (l/l0 - 1)^2 l0
    const strain = len / rest.edgeLengthsMm[j] - 1;
    stretchEnergy += 0.5 * material.axialStiffnessN * strain * strain * rest.edgeLengthsMm[j];
    const force = material.axialStiffnessN * strain; // [N]
    for (let k = 0; k < 3; k += 1) {
      const contribution = force * tangent[a + k];
      gradNodes[b + k] += contribution;
      gradNodes[a + k] -= contribution;
    }
  }

  // --- bending: E = EI / (2 lbar) |kb - kb0|^2
  const voronoi = voronoiLengths(rest);
  for (let i = 0; i < n - 2; i += 1) {
    const e0 = 3 * i;
    const e1 = 3 * (i + 1);
    const n0 = edgeLen[i];
    const n1 = edgeLen[i + 1];
    const dot =
      edgeVec[e0] * edgeVec[e1] +
      edgeVec[e0 + 1] * edgeVec[e1 + 1] +
      edgeVec[e0 + 2] * edgeVec[e1 + 2];
    const denominator = n0 * n1 + dot;
    if (!(denominator > EPSILON)) {
      throw new RodGeometryError(
        `Rod folded back on itself at node ${i + 1} (turning angle reached 180 deg). ` +
          'Reduce the time step or the steering input.',
      );
    }
    // kb = 2 e0 x e1 / denominator
    const cx = edgeVec[e0 + 1] * edgeVec[e1 + 2] - edgeVec[e0 + 2] * edgeVec[e1 + 1];
    const cy = edgeVec[e0 + 2] * edgeVec[e1] - edgeVec[e0] * edgeVec[e1 + 2];
    const cz = edgeVec[e0] * edgeVec[e1 + 1] - edgeVec[e0 + 1] * edgeVec[e1];
    const kbx = (2 * cx) / denominator;
    const kby = (2 * cy) / denominator;
    const kbz = (2 * cz) / denominator;

    const dkbx = kbx - rest.kb0[3 * i];
    const dkby = kby - rest.kb0[3 * i + 1];
    const dkbz = kbz - rest.kb0[3 * i + 2];

    const coefficient = material.bendingStiffnessNmm2 / voronoi[i]; // [N mm]
    bendEnergy += 0.5 * coefficient * (dkbx * dkbx + dkby * dkby + dkbz * dkbz);

    // w = dE/dkb
    const wx = coefficient * dkbx;
    const wy = coefficient * dkby;
    const wz = coefficient * dkbz;
    const wDotKb = wx * kbx + wy * kby + wz * kbz;

    // w^T dkb/de0 = (-2 (w x e1) - (w.kb) (n1 t0 + e1)) / denominator
    const wxe1x = wy * edgeVec[e1 + 2] - wz * edgeVec[e1 + 1];
    const wxe1y = wz * edgeVec[e1] - wx * edgeVec[e1 + 2];
    const wxe1z = wx * edgeVec[e1 + 1] - wy * edgeVec[e1];
    const a0x = (-2 * wxe1x - wDotKb * (n1 * tangent[e0] + edgeVec[e1])) / denominator;
    const a0y = (-2 * wxe1y - wDotKb * (n1 * tangent[e0 + 1] + edgeVec[e1 + 1])) / denominator;
    const a0z = (-2 * wxe1z - wDotKb * (n1 * tangent[e0 + 2] + edgeVec[e1 + 2])) / denominator;

    // w^T dkb/de1 = ( 2 (w x e0) - (w.kb) (n0 t1 + e0)) / denominator
    const wxe0x = wy * edgeVec[e0 + 2] - wz * edgeVec[e0 + 1];
    const wxe0y = wz * edgeVec[e0] - wx * edgeVec[e0 + 2];
    const wxe0z = wx * edgeVec[e0 + 1] - wy * edgeVec[e0];
    const a1x = (2 * wxe0x - wDotKb * (n0 * tangent[e1] + edgeVec[e0])) / denominator;
    const a1y = (2 * wxe0y - wDotKb * (n0 * tangent[e1 + 1] + edgeVec[e0 + 1])) / denominator;
    const a1z = (2 * wxe0z - wDotKb * (n0 * tangent[e1 + 2] + edgeVec[e0 + 2])) / denominator;

    const p0 = 3 * i;
    const p1 = 3 * (i + 1);
    const p2 = 3 * (i + 2);
    gradNodes[p0] -= a0x;
    gradNodes[p0 + 1] -= a0y;
    gradNodes[p0 + 2] -= a0z;
    gradNodes[p1] += a0x - a1x;
    gradNodes[p1 + 1] += a0y - a1y;
    gradNodes[p1 + 2] += a0z - a1z;
    gradNodes[p2] += a1x;
    gradNodes[p2 + 1] += a1y;
    gradNodes[p2 + 2] += a1z;

    // --- twist: E = GJ / (2 lbar) (dphi - dphi0)^2
    const twistIncrement = phi[i + 1] - phi[i] - rest.dphi0Rad[i];
    const twistCoefficient = material.torsionalStiffnessNmm2 / voronoi[i]; // [N mm]
    twistEnergy += 0.5 * twistCoefficient * twistIncrement * twistIncrement;
    const moment = twistCoefficient * twistIncrement;
    gradPhi[i + 1] += moment;
    gradPhi[i] -= moment;
  }

  // --- external loads
  if (loads?.nodeForcesN) {
    for (let index = 0; index < 3 * n; index += 1) {
      externalEnergy -= loads.nodeForcesN[index] * x[index];
      gradNodes[index] -= loads.nodeForcesN[index];
    }
  }
  if (loads?.edgeTorquesNmm) {
    for (let j = 0; j < edges; j += 1) {
      externalEnergy -= loads.edgeTorquesNmm[j] * phi[j];
      gradPhi[j] -= loads.edgeTorquesNmm[j];
    }
  }

  const total = stretchEnergy + bendEnergy + twistEnergy + externalEnergy;
  if (!Number.isFinite(total)) {
    throw new RodGeometryError(
      'Non-finite energy (NaN/Inf). The solver was stopped to avoid producing ' +
        'meaningless geometry. Check the parameter profile and the time step.',
    );
  }
  return { stretch: stretchEnergy, bend: bendEnergy, twist: twistEnergy, external: externalEnergy, total };
}

/** Build a straight, untwisted rod. Mirrors `straight_rod` in the Python core. */
export function straightRod(
  originMm: readonly number[],
  direction: readonly number[],
  totalLengthMm: number,
  nNodes: number,
  d1First?: readonly number[],
): { state: RodState; rest: RodRest } {
  if (nNodes < 3) throw new RodGeometryError('A discrete elastic rod needs at least 3 nodes.');
  if (!(totalLengthMm > 0)) throw new RodGeometryError('totalLengthMm must be positive.');
  const norm = Math.hypot(direction[0], direction[1], direction[2]);
  if (!(norm > EPSILON)) throw new RodGeometryError('direction must be a non-zero vector.');
  const unit = [direction[0] / norm, direction[1] / norm, direction[2] / norm];

  const nodesMm = new Float64Array(3 * nNodes);
  for (let i = 0; i < nNodes; i += 1) {
    const s = (totalLengthMm * i) / (nNodes - 1);
    nodesMm[3 * i] = originMm[0] + s * unit[0];
    nodesMm[3 * i + 1] = originMm[1] + s * unit[1];
    nodesMm[3 * i + 2] = originMm[2] + s * unit[2];
  }
  const edgeLengthsMm = new Float64Array(nNodes - 1).fill(totalLengthMm / (nNodes - 1));
  return {
    state: {
      nodesMm,
      phiRad: new Float64Array(nNodes - 1),
      d1First: Float64Array.from(d1First ?? anyPerpendicular(unit)),
    },
    rest: {
      edgeLengthsMm,
      kb0: new Float64Array(3 * (nNodes - 2)),
      dphi0Rad: new Float64Array(nNodes - 2),
    },
  };
}

/** An arbitrary unit vector perpendicular to `tangent`. */
export function anyPerpendicular(tangent: readonly number[]): number[] {
  const helper = Math.abs(tangent[2]) > 0.9 ? [1, 0, 0] : [0, 0, 1];
  const cx = tangent[1] * helper[2] - tangent[2] * helper[1];
  const cy = tangent[2] * helper[0] - tangent[0] * helper[2];
  const cz = tangent[0] * helper[1] - tangent[1] * helper[0];
  const norm = Math.hypot(cx, cy, cz);
  return [cx / norm, cy / norm, cz / norm];
}

File: packages/physics-core/src/cathsim_physics/rod.py
"""Discrete elastic rod - reference implementation (Python + NumPy).

Continuous model
----------------
The catheter shaft is a Kirchhoff / Cosserat rod whose stored elastic energy is

.. code-block:: text

    E = integral [ 1/2 (kappa - kappa0)^T B (kappa - kappa0)
                 + 1/2 (tau   - tau0  )^T C (tau   - tau0  )
                 + 1/2 (eps   - eps0  )^T S (eps   - eps0  ) ] ds

with bending stiffness ``B = EI`` [N mm^2], torsional stiffness ``C = GJ``
[N mm^2] and axial stiffness ``S = EA`` [N].  ``docs/physics-model.md`` derives
the discretisation below from this integral term by term.

Discretisation (Bishop-frame discrete elastic rod)
--------------------------------------------------
* ``N`` nodes ``x_0 .. x_{N-1}`` [mm], ``M = N-1`` edges ``e_j = x_{j+1}-x_j``.
* One scalar material twist angle ``phi_j`` [rad] per edge, measured against a
  Bishop (rotation-minimising, twist-free) frame anchored on edge 0.
* Discrete curvature binormal at interior node ``i``:
  ``kb_i = 2 e_{i-1} x e_i / (|e_{i-1}||e_i| + e_{i-1}.e_i)``  [-]
  which satisfies ``|kb_i| = 2 tan(phi_i/2) ~= curvature_i * lbar_i``.
* Voronoi length ``lbar_i = (l_{i-1} + l_i)/2`` [mm].

Discrete energies (all in N mm):

.. code-block:: text

    E_stretch = sum_j  1/2 EA (|e_j|/l_j - 1)^2 l_j
    E_bend    = sum_i  EI / (2 lbar_i) |kb_i - kb0_i|^2
    E_twist   = sum_i  GJ / (2 lbar_i) (phi_i - phi_{i-1} - dphi0_i)^2

``kb0_i`` is the **rest curvature binormal expressed in world coordinates**.
Steering writes into it (see :mod:`cathsim_physics.catheter`).  It is held
fixed while the solver relaxes one actuation state and refreshed from the
material frame at the start of the next step; at equilibrium this is a fixed
point, so the converged shape is consistent.

Modelling scope of this MVP (see ``docs/assumptions.md``):

* Bending is **isotropic** (``B1 = B2 = EI``), correct for the round
  cross-section assumed here.  Anisotropic bending is future work.
* Because the twist DOF is measured against a twist-free Bishop frame,
  ``E_twist`` has no explicit position dependence, so ``dE_twist/dx = 0``
  *exactly* in this parameterisation.  The bend-twist (writhe/holonomy)
  coupling therefore only enters through a *twist-prescribed* boundary.  The
  catheter is clamped proximally and free distally, so the omitted coupling
  term does not change the equilibrium.  This is documented, not accidental.
* Shear is suppressed by the inextensible-direction assumption of the
  Kirchhoff rod (tangent == material director 3); no separate shear DOF.

Research prototype - Not for clinical use.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from .frames import bishop_frame, material_frame, normalise

_EPS = 1.0e-12


@dataclass(frozen=True)
class RodMaterial:
    """Constitutive parameters of the rod, all in the project unit system.

    Attributes:
        bending_stiffness_n_mm2: ``EI`` [N mm^2].
        torsional_stiffness_n_mm2: ``GJ`` [N mm^2].
        axial_stiffness_n: ``EA`` [N].
        damping_n_s_per_mm: lumped nodal translational viscous damping
            coefficient.  NOTE: the *key name* is fixed by the project
            contract; the coefficient is applied **per unit length**, so its
            effective unit is [N s / mm^2].  See ``docs/assumptions.md``.
        torsional_damping_n_mm_s_per_rad: viscous damping on the twist DOF
            [N mm s / rad], applied per unit length.
        linear_density_kg_per_mm: mass per unit length, only used when gravity
            is enabled.  ``None`` means gravity cannot be enabled.
    """

    bending_stiffness_n_mm2: float
    torsional_stiffness_n_mm2: float
    axial_stiffness_n: float
    damping_n_s_per_mm: float
    torsional_damping_n_mm_s_per_rad: float
    linear_density_kg_per_mm: float | None = None


@dataclass
class RodState:
    """Kinematic state of the rod.

    Attributes:
        nodes_mm: ``(N, 3)`` node positions [mm].
        phi_rad: ``(N-1,)`` material twist angle per edge [rad].
        d1_first: ``(3,)`` Bishop director anchor of edge 0 [-].
    """

    nodes_mm: np.ndarray
    phi_rad: np.ndarray
    d1_first: np.ndarray

    def copy(self) -> RodState:
        """Return an independent deep copy of this state."""
        return RodState(self.nodes_mm.copy(), self.phi_rad.copy(), self.d1_first.copy())

    @property
    def n_nodes(self) -> int:
        """Number of nodes N."""
        return int(self.nodes_mm.shape[0])

    @property
    def n_edges(self) -> int:
        """Number of edges M = N - 1."""
        return int(self.nodes_mm.shape[0] - 1)


@dataclass
class RodRest:
    """Undeformed (rest) configuration of the rod.

    Attributes:
        edge_lengths_mm: ``(M,)`` rest length of every edge [mm].
        kb0: ``(N-2, 3)`` rest curvature binormal at interior nodes [-].
        dphi0_rad: ``(N-2,)`` rest twist increment per interior node [rad].
    """

    edge_lengths_mm: np.ndarray
    kb0: np.ndarray
    dphi0_rad: np.ndarray

    @property
    def voronoi_lengths_mm(self) -> np.ndarray:
        """``(N-2,)`` Voronoi length of every interior node [mm]."""
        return 0.5 * (self.edge_lengths_mm[:-1] + self.edge_lengths_mm[1:])


@dataclass
class ExternalLoads:
    """External generalised loads. Empty by default.

    Attributes:
        node_forces_n: ``(N, 3)`` force applied to each node [N].
        edge_torques_n_mm: ``(M,)`` twisting moment applied to each edge's twist
            DOF [N mm].
        gravity_mm_per_s2: ``(3,)`` gravity vector [mm/s^2], or ``None`` for off.
    """

    node_forces_n: np.ndarray | None = None
    edge_torques_n_mm: np.ndarray | None = None
    gravity_mm_per_s2: np.ndarray | None = None


@dataclass
class EnergyBreakdown:
    """Elastic energy split by mode [N mm]."""

    stretch: float = 0.0
    bend: float = 0.0
    twist: float = 0.0
    external: float = 0.0

    @property
    def elastic_total(self) -> float:
        """Stretch + bend + twist energy [N mm]."""
        return self.stretch + self.bend + self.twist

    @property
    def total(self) -> float:
        """Elastic energy plus the external-load potential [N mm]."""
        return self.elastic_total + self.external


def curvature_binormal(nodes_mm: np.ndarray) -> np.ndarray:
    """Discrete curvature binormal ``kb`` at every interior node.

    Args:
        nodes_mm: ``(N, 3)`` node positions [mm].

    Returns:
        ``(N-2, 3)`` dimensionless curvature binormals.
    """
    e0 = nodes_mm[1:-1] - nodes_mm[:-2]
    e1 = nodes_mm[2:] - nodes_mm[1:-1]
    n0 = np.linalg.norm(e0, axis=1)
    n1 = np.linalg.norm(e1, axis=1)
    denom = n0 * n1 + np.sum(e0 * e1, axis=1)
    if np.any(denom < _EPS):
        raise ValueError(
            "Rod folded back on itself (a discrete turning angle reached 180 deg). "
            "Reduce the time step, lower the steering input, or refine the node "
            "spacing."
        )
    return 2.0 * np.cross(e0, e1) / denom[:, None]


def _check_finite(name: str, array: np.ndarray) -> None:
    if not np.all(np.isfinite(array)):
        raise ValueError(
            f"Non-finite value (NaN/Inf) encountered in '{name}'. The solver was "
            f"stopped to avoid producing meaningless geometry. Check the input "
            f"configuration and the time step."
        )


def energy_and_gradient(
    state: RodState,
    rest: RodRest,
    material: RodMaterial,
    loads: ExternalLoads | None = None,
) -> tuple[EnergyBreakdown, np.ndarray, np.ndarray]:
    """Total energy [N mm] and its gradient w.r.t. every degree of freedom.

    Returns:
        ``(energy, grad_nodes, grad_phi)`` where ``grad_nodes`` is ``(N, 3)``
        in newtons and ``grad_phi`` is ``(M,)`` in newton-millimetres.
    """
    nodes = state.nodes_mm
    phi = state.phi_rad
    _check_finite("nodes_mm", nodes)
    _check_finite("phi_rad", phi)

    grad_nodes = np.zeros_like(nodes)
    grad_phi = np.zeros_like(phi)
    energy = EnergyBreakdown()

    # ---- axial stretch -------------------------------------------------
    edges = nodes[1:] - nodes[:-1]
    lengths = np.linalg.norm(edges, axis=1)
    if np.any(lengths < _EPS):
        raise ValueError(
            "Two neighbouring rod nodes collapsed (zero edge length). Increase "
            "the axial stiffness EA or reduce the time step."
        )
    strain = lengths / rest.edge_lengths_mm - 1.0
    energy.stretch = float(
        0.5 * material.axial_stiffness_n * np.sum(strain**2 * rest.edge_lengths_mm)
    )
    tangents = edges / lengths[:, None]
    edge_force = (material.axial_stiffness_n * strain)[:, None] * tangents  # [N]
    grad_nodes[1:] += edge_force
    grad_nodes[:-1] -= edge_force

    # ---- bending -------------------------------------------------------
    voronoi = rest.voronoi_lengths_mm
    e0 = nodes[1:-1] - nodes[:-2]
    e1 = nodes[2:] - nodes[1:-1]
    n0 = lengths[:-1]
    n1 = lengths[1:]
    denom = n0 * n1 + np.sum(e0 * e1, axis=1)
    if np.any(denom < _EPS):
        raise ValueError(
            "Rod folded back on itself (a discrete turning angle reached 180 deg). "
            "Reduce the time step or the steering input."
        )
    kb = 2.0 * np.cross(e0, e1) / denom[:, None]
    delta_kb = kb - rest.kb0
    bend_coeff = material.bending_stiffness_n_mm2 / voronoi  # [N mm]
    energy.bend = float(0.5 * np.sum(bend_coeff * np.sum(delta_kb**2, axis=1)))

    # w = dE/dkb  (N mm, dimensionless kb -> newton-millimetre)
    w = bend_coeff[:, None] * delta_kb
    # Using  w^T [v]_x = w x v :
    #   w^T dkb/de0 = (-2 (w x e1) - (w.kb) (n1 t0 + e1)) / denom
    #   w^T dkb/de1 = ( 2 (w x e0) - (w.kb) (n0 t1 + e0)) / denom
    w_dot_kb = np.sum(w * kb, axis=1)[:, None]
    t0 = tangents[:-1]
    t1 = tangents[1:]
    a0 = (-2.0 * np.cross(w, e1) - w_dot_kb * (n1[:, None] * t0 + e1)) / denom[:, None]
    a1 = (2.0 * np.cross(w, e0) - w_dot_kb * (n0[:, None] * t1 + e0)) / denom[:, None]
    grad_nodes[:-2] -= a0
    grad_nodes[1:-1] += a0 - a1
    grad_nodes[2:] += a1

    # ---- twist ---------------------------------------------------------
    twist_increment = phi[1:] - phi[:-1] - rest.dphi0_rad
    twist_coeff = material.torsional_stiffness_n_mm2 / voronoi  # [N mm]
    energy.twist = float(0.5 * np.sum(twist_coeff * twist_increment**2))
    moment = twist_coeff * twist_increment  # [N mm]
    grad_phi[1:] += moment
    grad_phi[:-1] -= moment

    # ---- external loads ------------------------------------------------
    if loads is not None:
        if loads.node_forces_n is not None:
            energy.external += float(-np.sum(loads.node_forces_n * nodes))
            grad_nodes -= loads.node_forces_n
        if loads.edge_torques_n_mm is not None:
            energy.external += float(-np.sum(loads.edge_torques_n_mm * phi))
            grad_phi -= loads.edge_torques_n_mm
        if loads.gravity_mm_per_s2 is not None:
            if material.linear_density_kg_per_mm is None:
                raise ValueError(
                    "Gravity was enabled but 'linear_density_kg_per_mm' is not "
                    "calibrated (CALIBRATION_REQUIRED). Disable gravity or supply "
                    "a density in the parameter profile."
                )
            # Lumped nodal mass from the Voronoi length of each node [kg].
            node_mass = np.zeros(state.n_nodes)
            half = 0.5 * rest.edge_lengths_mm * material.linear_density_kg_per_mm
            node_mass[:-1] += half
            node_mass[1:] += half
            # 1 kg * 1 mm/s^2 = 1e-3 N  ->  weight [N] = m[kg]*g[mm/s^2]*1e-3
            weight = node_mass[:, None] * loads.gravity_mm_per_s2[None, :] * 1.0e-3
            energy.external += float(-np.sum(weight * nodes))
            grad_nodes -= weight

    _check_finite("energy gradient (nodes)", grad_nodes)
    _check_finite("energy gradient (twist)", grad_phi)
    return energy, grad_nodes, grad_phi


def bishop_directors(state: RodState) -> tuple[np.ndarray, np.ndarray]:
    """Return ``(tangents, d1)`` for the current state, both ``(M, 3)``."""
    edges = state.nodes_mm[1:] - state.nodes_mm[:-1]
    tangents = normalise(edges)
    d1 = bishop_frame(tangents, state.d1_first)
    return tangents, d1


def material_directors(state: RodState) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Return ``(tangents, m1, m2)`` of the material frame, each ``(M, 3)``."""
    tangents, d1 = bishop_directors(state)
    m1, m2 = material_frame(tangents, d1, state.phi_rad)
    return tangents, m1, m2


def straight_rod(
    origin_mm: np.ndarray,
    direction: np.ndarray,
    total_length_mm: float,
    n_nodes: int,
    d1_first: np.ndarray | None = None,
) -> tuple[RodState, RodRest]:
    """Build a straight, untwisted rod of ``n_nodes`` nodes.

    Args:
        origin_mm: ``(3,)`` position of node 0 [mm].
        direction: ``(3,)`` direction the rod points in (need not be unit).
        total_length_mm: arclength from node 0 to node N-1 [mm].
        n_nodes: number of nodes, must be >= 3.
        d1_first: optional Bishop anchor director; an arbitrary perpendicular
            is chosen when omitted.
    """
    if n_nodes < 3:
        raise ValueError("A discrete elastic rod needs at least 3 nodes.")
    if total_length_mm <= 0.0:
        raise ValueError("total_length_mm must be positive.")
    unit = normalise(np.asarray(direction, dtype=float)[None, :])[0]
    s = np.linspace(0.0, total_length_mm, n_nodes)
    nodes = np.asarray(origin_mm, dtype=float)[None, :] + s[:, None] * unit[None, :]
    edge_lengths = np.full(n_nodes - 1, total_length_mm / (n_nodes - 1))
    if d1_first is None:
        from .frames import any_perpendicular

        d1_first = any_perpendicular(unit)
    state = RodState(
        nodes_mm=nodes,
        phi_rad=np.zeros(n_nodes - 1),
        d1_first=np.asarray(d1_first, dtype=float),
    )
    rest = RodRest(
        edge_lengths_mm=edge_lengths,
        kb0=np.zeros((n_nodes - 2, 3)),
        dphi0_rad=np.zeros(n_nodes - 2),
    )
    return state, rest
